force local wc_nnn hash struct variables to zeros

gojimmypi opened the pull request:
# Description

Related to the https://github.com/wolfSSL/wolfssl/issues/5948 ED25519 failure, this update forces the local SHA struct values to all zeros before using them. This is typically only important to hash hardware acceleration where existing HW state may be stored in the `ctx` and otherwise have potentially bad data if not properly initialized or worse: copied from another SHA `ctx` that may have active hardware acceleration computation state values.

Recall only one given HW acceleration computation can be in process at a given time on the ESP32. Although the ESP32-C3 does seem to have the ability for concurrent HW encryption, that's beyond the scope of this PR and in any case we'd still need a properly initialized variable anyhow.

Note this PR is *not* a substitute for proper SHA ctx self validation to check for creation from a copy as noted in https://github.com/wolfSSL/wolfssl/issues/5948#issuecomment-1421288725, and is not meant to be a final fix to https://github.com/wolfSSL/wolfssl/issues/5948 .  A separate PR will be created soon.

There's a small performance hit when assigning zeros at every use. I'm open to suggestions if this should only be applied for relevant platforms, such as the Espressif ESP32 with Hardware Acceleration. I chose to do it all the time as a Best Practice to have variables always safely initialized.

Fixes zd# n/a

# Testing

How did you test? Ran the wolfcrypt test for both embedded ESP32 and Linux.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1219,6 +1219,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the md5 we start with does not have
+        ** any unexpected values in any of the properties. */
+        ForceZero(md5, sizeof(wc_Md5));
```
Just use `XMEMSET` here please. The ForceZero is possible not as fast and is only required in places where there might be secrets that need cleared after use. This is just ensuring the data is zero'd, there shouldn't be any secrets left behind. Also this now duplicates work already done in some of the init functions. If using memset the compiler could optimize away the duplicate (not if ForceZero is used).

dgarske replied:
This will have performance implications. My preference is to fix it in `wc_InitMd5`.

gojimmypi replied:
I've reverted the changes to `hash.c` and moved the memory zeroize, now with `XMEMSET` instead of `ForceZero` into the respective `init`s.

gojimmypi replied:
> duplicates work already done in some of the init functions

@dgarske regarding this topic,: Although the ESP32 SHA tests pass, the RSA fails when the apparently duplicate `XMEMSET` [here](https://github.com/gojimmypi/wolfssl/blob/6a17a140076333f7bff226f851133e6ba301a9f9/wolfcrypt/src/sha256.c#L778) which _should_ be ok to omit, as it is also [here](https://github.com/gojimmypi/wolfssl/blob/6a17a140076333f7bff226f851133e6ba301a9f9/wolfcrypt/src/sha256.c#L738) in the call to `InitSha256`.  Clearly there's more research to be done.

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1257,6 +1262,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the sha we start with does not have
+        ** any unexpected values in any of the properties */
+        ForceZero(sha, sizeof(wc_Sha));
```
Same

gojimmypi replied:
moved to `md5.c`

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1301,6 +1311,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the sha224 we start with does not have
+        ** any unexpected values in any of the properties */
+        ForceZero(sha224, sizeof(wc_Sha224));
```
Same

gojimmypi replied:
moved to `sha256.c`

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1340,6 +1355,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the sha256 we start with does not have
+        ** any unexpected values in any of the properties */
+        ForceZero(sha256, sizeof(wc_Sha256));
```
Same

gojimmypi replied:
moved to `sha256.c`

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1388,6 +1408,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the sha512 we start with does not have
+        ** any unexpected values in any of the properties. */
+        ForceZero(sha512, sizeof(wc_Sha512));
```
Same

gojimmypi replied:
moved to `sha512.c`

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1425,6 +1450,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the sha512 we start with does not have
+        ** any unexpected values in any of the properties. */
+        ForceZero(sha512, sizeof(wc_Sha512));
```
Same

gojimmypi replied:
moved to `sha512.c`

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1465,6 +1495,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the sha we start with does not have
+        ** any unexpected values in any of the properties. */
+        ForceZero(sha512, sizeof(wc_Sha512));
```
Same

gojimmypi replied:
moved to `sha512.c`

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1506,6 +1541,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the sha384 we start with does not have
+        ** any unexpected values in any of the properties. */
+        ForceZero(sha384, sizeof(wc_Sha384));
```
Same

gojimmypi replied:
moved to `sha512.c`

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1545,6 +1585,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the sha3 we start with does not have
+        ** any unexpected values in any of the properties. */
+        ForceZero(sha3, sizeof(wc_Sha3));
```
Same

gojimmypi replied:
moved to `sha3c`

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1583,6 +1628,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the sha3 we start with does not have
+        ** any unexpected values in any of the properties. */
+        ForceZero(sha3, sizeof(wc_Sha3));
```
Same

gojimmypi replied:
moved to `sha3c`

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1621,6 +1671,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the sha3 we start with does not have
+        ** any unexpected values in any of the properties. */
+        ForceZero(sha3, sizeof(wc_Sha3));
```
Same

gojimmypi replied:
moved to `sha3c`

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1659,6 +1714,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the sha3 we start with does not have
+        ** any unexpected values in any of the properties. */
+        ForceZero(sha3, sizeof(wc_Sha3));
```
Same

gojimmypi replied:
moved to `sha3c`

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1698,6 +1758,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the shake we start with does not have
+        ** any unexpected values in any of the properties. */
+        ForceZero(shake, sizeof(wc_Shake));
```
Same

gojimmypi replied:
moved to `sha3c`

dgarske commented on wolfcrypt/src/hash.c:
```diff
@@ -1737,6 +1802,11 @@ int wc_HashGetFlags(wc_HashAlg* hash, enum wc_HashType type, word32* flags)
+        /* Typically only important to hardware acceleration, but a good
+        ** practice: we'll make sure the shake we start with does not have
+        ** any unexpected values in any of the properties. */
+        ForceZero(shake, sizeof(wc_Shake));
```
Same

gojimmypi replied:
moved to `sha3c`

dgarske reviewed (changes requested):
Jim please consider properly fixing the initialization it in each of the hash init routines. Some of these already do a memset and this is now duplicated. Also the ForceZero should be memset.

gojimmypi commented:
Reverted `hash.c` and moved `XMEMSET` into various files for initializtions. I'll squash merge after review if desired.

Rechecked with wolftest:

```
make[2]: Leaving directory '/mnt/c/temp/wolfssl_PR6068_test'
make[1]: Leaving directory '/mnt/c/temp/wolfssl_PR6068_test'
------------------------------------------------------------------------------
 wolfSSL version 5.5.4
------------------------------------------------------------------------------
error    test passed!
MEMORY   test passed!
base64   test passed!
asn      test passed!
RANDOM   test passed!
MD5      test passed!
SHA      test passed!
SHA-224  test passed!
SHA-256  test passed!
SHA-384  test passed!
SHA-512  test passed!
SHA-3    test passed!
Hash     test passed!
HMAC-MD5 test passed!
HMAC-SHA test passed!
HMAC-SHA224 test passed!
HMAC-SHA256 test passed!
HMAC-SHA384 test passed!
HMAC-SHA512 test passed!
HMAC-SHA3   test passed!
HMAC-KDF    test passed!
TLSv1.3 KDF test passed!
GMAC     test passed!
Chacha   test passed!
POLY1305 test passed!
ChaCha20-Poly1305 AEAD test passed!
AES      test passed!
AES192   test passed!
AES256   test passed!
AES-GCM  test passed!
RSA      test passed!
DH       test passed!
PWDBASED test passed!
ECC      test passed!
logging  test passed!
time test passed!
mutex    test passed!
memcb    test passed!
Test complete
stack used = 2097152
total   Allocs   =      1079
total   Deallocs =      1079
total   Bytes    =   2948064
peak    Bytes    =     59930
current Bytes    =         0
Exiting main with return code: 0
```

dgarske commented on wolfcrypt/src/sha512.c:
```diff
@@ -1618,8 +1622,14 @@ int wc_Sha512GetFlags(wc_Sha512* sha512, word32* flags)
 #if !defined(WOLFSSL_NOSHA512_224)
 int wc_InitSha512_224(wc_Sha512* sha)
 {
+    if (sha == NULL)
```
This is redundant. Already done in `wc_InitSha512_224_ex` -> `InitSha512_Family`.

gojimmypi replied:
agreed. good catch. updated.

dgarske reviewed (approved):
Thanks Jim. FYI: I will squash this on merge.

SparkiDev commented on wolfcrypt/src/sha256.c:
```diff
@@ -1567,17 +1567,21 @@ static int InitSha256(wc_Sha256* sha256)
             return BAD_FUNC_ARG;
+        }
+
+        XMEMSET(sha224, 0, sizeof(wc_Sha224));
```
Missed wc_InitSha256_ex()

gojimmypi replied:
@SparkiDev  thanks for your review! Would you recommend that I do a `XMEMSET` in each of these?

![image](https://user-images.githubusercontent.com/13059545/218228929-2bbf2f73-2c67-425a-acc0-80d15031b456.png)

I'm not entirely convinced it is a good idea for software-only hashes. The only reason I found this and started to do the initializations was due to undesired hardware encryption state data being _copied_ into a new sha `ctx` instance. 

Please advise. Thank you.

SparkiDev replied:
If the PR is to memset the data structure when using hardware and they are the implementations for hardware, then yes.
If you can't find a common for each function, then it needs to be done in each.

SparkiDev commented on wolfcrypt/src/sha512.c:
```diff
@@ -573,10 +573,13 @@ static int InitSha512_256(wc_Sha512* sha512)
         return BAD_FUNC_ARG;
+    }
+
+    XMEMSET(sha512, 0, sizeof(wc_Sha512));
```
Missed wc_InitSha383_ex().

gojimmypi commented:
I have a higher performance, less brute-force method of initialization in the works. Closing this PR; New PR will be created from my [WIP ED25519_SHA2_fix Branch](https://github.com/gojimmypi/wolfssl/tree/ED25519_SHA2_fix).

gojimmypi commented:
For reference, the root cause of the ED25519 failures are noted in https://github.com/wolfSSL/wolfssl/issues/5948#issuecomment-1533240229 and fixed in https://github.com/wolfSSL/wolfssl/pull/6287.

The root cause was the use of in-progress `ctx` _copies_ during active SHA HW process.
